# Python Insights - Analyzing Data with Python

### Case - Customer Cancellation

You have been hired by a company with over 800,000 customers for a Data project. The company recently realized that of its total customer base, the majority are inactive customers, that is, those who have already canceled the service.

In order to improve its results, the company wants to understand the main reasons for these cancellations and what are the most efficient actions to reduce this number.

## Importing libraries

In [2]:
import pandas as pd
import plotly.express as px

## Loading the database and reading our data frame

In [5]:
df = pd.read_csv('cancelamentos.csv')
df

CustomerID  idade    sexo  tempo_como_cliente  frequencia_uso  \
0              2.0   30.0  Female                39.0            14.0   
1              3.0   65.0  Female                49.0             1.0   
2              4.0   55.0  Female                14.0             4.0   
3              5.0   58.0    Male                38.0            21.0   
4              6.0   23.0    Male                32.0            20.0   
...            ...    ...     ...                 ...             ...   
881661    449995.0   42.0    Male                54.0            15.0   
881662    449996.0   25.0  Female                 8.0            13.0   
881663    449997.0   26.0    Male                35.0            27.0   
881664    449998.0   28.0    Male                55.0            14.0   
881665    449999.0   31.0    Male                48.0            20.0   

        ligacoes_callcenter  dias_atraso assinatura duracao_contrato  \
0                       5.0         18.0   Standard           Annual   
1                      10.0          8.0      Basic          Monthly   
2                       6.0         18.0      Basic        Quarterly   
3                       7.0          7.0   Standard          Monthly   
4                       5.0          8.0      Basic          Monthly   
...                     ...          ...        ...              ...   
881661                  1.0          3.0    Premium           Annual   
881662                  1.0         20.0    Premium           Annual   
881663                  1.0          5.0   Standard        Quarterly   
881664                  2.0          0.0   Standard        Quarterly   
881665                  1.0         14.0    Premium        Quarterly   

        total_gasto  meses_ultima_interacao  cancelou  
0            932.00                    17.0       1.0  
1            557.00                     6.0       1.0  
2            185.00                     3.0       1.0  
3            396.00                    29.0       1.0  
4            617.00                    20.0       1.0  
...             ...                     ...       ...  
881661       716.38                     8.0       0.0  
881662       745.38                     2.0       0.0  
881663       977.31                     9.0       0.0  
881664       602.55                     2.0       0.0  
881665       567.77                    21.0       0.0  

[881666 rows x 12 columns]

## Cleaning and filtering data

### Looking for null values
Let's see if we can find any inconsistency just with the info method.

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 881666 entries, 0 to 881665
Data columns (total 12 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   CustomerID              881664 non-null  float64
 1   idade                   881664 non-null  float64
 2   sexo                    881664 non-null  object 
 3   tempo_como_cliente      881663 non-null  float64
 4   frequencia_uso          881663 non-null  float64
 5   ligacoes_callcenter     881664 non-null  float64
 6   dias_atraso             881664 non-null  float64
 7   assinatura              881661 non-null  object 
 8   duracao_contrato        881663 non-null  object 
 9   total_gasto             881664 non-null  float64
 10  meses_ultima_interacao  881664 non-null  float64
 11  cancelou                881664 non-null  float64
dtypes: float64(9), object(3)
memory usage: 80.7+ MB



Well, it looks like there are missing values on some columns. Let's look closer.

In [12]:
null_df = df[df.isnull().any(axis=1)]
null_df

CustomerID  idade    sexo  tempo_como_cliente  frequencia_uso  \
11            14.0   52.0  Female                21.0             6.0   
14            17.0   24.0    Male                 4.0             9.0   
17            20.0   47.0    Male                41.0             NaN   
18            21.0   24.0    Male                44.0            13.0   
23            26.0   27.0  Female                 NaN             8.0   
199295         NaN    NaN     NaN                 NaN             NaN   
640128         NaN    NaN     NaN                 NaN             NaN   

        ligacoes_callcenter  dias_atraso assinatura duracao_contrato  \
11                      3.0         26.0    Premium              NaN   
14                      4.0         22.0        NaN        Quarterly   
17                      1.0          5.0        NaN           Annual   
18                      5.0          4.0        NaN          Monthly   
23                      7.0          3.0   Standard          Monthly   
199295                  NaN          NaN        NaN              NaN   
640128                  NaN          NaN        NaN              NaN   

        total_gasto  meses_ultima_interacao  cancelou  
11            830.0                    19.0       1.0  
14            204.0                     4.0       1.0  
17            151.0                    19.0       1.0  
18            669.0                    13.0       1.0  
23            434.0                    19.0       1.0  
199295          NaN                     NaN       NaN  
640128          NaN                     NaN       NaN

So, we found some missing information on seven rows, from a total of 881666 rows.
It won't hurt if we discard these. Let's clean this little mess!

### Getting rid of null values

In [16]:
df = df.dropna()

### Filtering by need

Now that the bad info has gone, we can proceed with our analysis.
But before we do it, as we are cleaning data, let's also discard another useless information, such as columns that certainly won't help with our problem.
Some columns may or may not have a correlation with our problem, we don't know yet. But the ID column has no relation with the cancellation. 
Dropping it will make the data clearer to see and lighter to proccess.

In [20]:
df = df.drop(columns=['CustomerID'])

In [22]:
df

idade    sexo  tempo_como_cliente  frequencia_uso  \
0        30.0  Female                39.0            14.0   
1        65.0  Female                49.0             1.0   
2        55.0  Female                14.0             4.0   
3        58.0    Male                38.0            21.0   
4        23.0    Male                32.0            20.0   
...       ...     ...                 ...             ...   
881661   42.0    Male                54.0            15.0   
881662   25.0  Female                 8.0            13.0   
881663   26.0    Male                35.0            27.0   
881664   28.0    Male                55.0            14.0   
881665   31.0    Male                48.0            20.0   

        ligacoes_callcenter  dias_atraso assinatura duracao_contrato  \
0                       5.0         18.0   Standard           Annual   
1                      10.0          8.0      Basic          Monthly   
2                       6.0         18.0      Basic        Quarterly   
3                       7.0          7.0   Standard          Monthly   
4                       5.0          8.0      Basic          Monthly   
...                     ...          ...        ...              ...   
881661                  1.0          3.0    Premium           Annual   
881662                  1.0         20.0    Premium           Annual   
881663                  1.0          5.0   Standard        Quarterly   
881664                  2.0          0.0   Standard        Quarterly   
881665                  1.0         14.0    Premium        Quarterly   

        total_gasto  meses_ultima_interacao  cancelou  
0            932.00                    17.0       1.0  
1            557.00                     6.0       1.0  
2            185.00                     3.0       1.0  
3            396.00                    29.0       1.0  
4            617.00                    20.0       1.0  
...             ...                     ...       ...  
881661       716.38                     8.0       0.0  
881662       745.38                     2.0       0.0  
881663       977.31                     9.0       0.0  
881664       602.55                     2.0       0.0  
881665       567.77                    21.0       0.0  

[881659 rows x 11 columns]

## Time to explore the data!

### First things first, we were told that the cancellations rate has increased. But how much? Is the situation even that bad? Time to find it out

In [24]:
count = df['cancelou'].value_counts()
display(count)

cancel_fig = px.pie(df, names='cancelou', title='Cancellation distribution')
cancel_fig.show()

cancelou
1.0    499993
0.0    381666
Name: count, dtype: int64

![Gráfico estático](plots/initial_rate.png)

499993 people have canceled the services, it's more than half of our client base! The situation isn't that bad, it is even worse!
But, let's give it a thought. There must be something really wrong happening on our data that justify the high cancellation rate. It can't be just coincidance... Can it?

### Time to compare every variable to the cancellation and see if we can find any correlation

In [31]:
for column in df.columns:
    fig = px.histogram(df, x=column, color='cancelou', text_auto=True)
    fig.show()

![Chart](plots/age.png)
![Chart](plots/gender.png)
![Chart](plots/time_as_customer.png)
![Chart](plots/callcenter.png)
![Chart](plots/frequency.png)
![Chart](plots/late.png)
![Chart](plots/type.png)
![Chart](plots/plan.png)
![Chart](plots/total_spent.png)
![Chart](plots/last.png)


At first glance, there are three variables that lead to discrepant results: the amount of calls from callcenter, the planning and the days late payment. Lets look closer at each one.

### First, the callcenter
It seems that once a customer receive more than five calls, he's very likely to want to cancel the service.

In [35]:
fig = px.histogram(df, x='ligacoes_callcenter', color='cancelou', text_auto=True)
fig.show()

![Callcenter chart](plots/callcenter.png)

Let's make an imaginary exercise and see what the results would be if the callcenter didn't exceeed the amount of five calls per customer. Then they wouldn't be too upset to leave us, right?

In [38]:
#Let's assume the callcenter wont' disturb customers anymore
new_df = df[df['ligacoes_callcenter'] < 5]

#Retry the counting
display(new_df['cancelou'].value_counts())
cancel_fig = px.pie(new_df, names='cancelou', title='Cancellation distribution')
cancel_fig.show()

cancelou
0.0    379032
1.0    215767
Name: count, dtype: int64

![Rate without callcenter calls](plots/callcenter_rate.png)

Wow, just by reducing the amount of inconvenient calls, we could almost reduce 10% of the cancellations. Who would have thought?
But our work is not already finished. There are two more variables to fix.

### Plan and late payment

In [41]:
fig1 = px.histogram(df, x='duracao_contrato', color='cancelou', title='Cancellation x Planning', text_auto=True)
fig1.show()
fig2 = px.histogram(df, x='dias_atraso', color='cancelou', title='Cancellation x Payment late', text_auto=True)
fig2.show()

![Plan chart](plots/plan.png)
![Late payment chart](plots/late.png)

Once again, we can find that in the first case, the monthly plan has 100% of cancellation. None of our customers that paid monthly stayed with us. If we could make people migrate their plan, it would be 174k cancellations converted, a little less than 40% of our initial 500k cancellation.
In the second case, people that takes more than twenty days to pay are high likely to cancel our services.

In [44]:
# Excluding the monthly plan
new_df = new_df[new_df['duracao_contrato'] != 'Monthly']

# Excluding payments with more than 20 days
new_df = new_df[new_df['dias_atraso'] <= 20]

### Other variables

There are other variables that interfere in the results, such as gender and age.
66% of women are likely to cancel, while for men it's about 50%. People past 50 years old are also certainly to cancel. But in these cases, there are nothing to do, as we're talking more about personal and subjective factors and less about the quality of our services. We can't simply not attend people older than 50 or women. Let's stick with the variables we gathered and call it a day.

## Testing the differences 

As stated before, we started with 499993 cancellations (or 56.8%). Let's see how much those management strategies would affect the current situation.

In [49]:
display(new_df['cancelou'].value_counts())

cancel_fig = px.pie(new_df, names='cancelou', title='Cancellation distribution')
cancel_fig.show()

cancelou
0.0    379032
1.0     85447
Name: count, dtype: int64

![Final rate](plots/final_rate.png)

 Amazing! We brought the cancellation rate from 56.8% to astonishing 18.4%, from 499993 to 85447. Just by changing three variables, we could reduce 67% of the cancellations! Of course, we'll never reach the 0% mark, but compared to before, the changes make the situation now looks like night and day!